File: 1926.迷宫中离入口最近的出口.go
/*
 * @lc app=leetcode.cn id=1926 lang=golang
 *
 * [1926] 迷宫中离入口最近的出口
 */

// @lc code=start
func nearestExit(maze [][]byte, entrance []int) int {
	// 直接 bfs
	m, n := len(maze), len(maze[0])
	directions := [4][2]int{{1, 0}, {-1, 0}, {0, 1}, {0, -1}}
	q := [][2]int{{entrance[0], entrance[1]}}

	// 判定是否越界
	inArea := func(x, y int) bool {
		if 0 <= x && x < m && 0 <= y && y < n {
			return true
		}
		return false
	}
	maze[entrance[0]][entrance[1]] = '+'
	for res := 1; len(q) > 0; res++ {
		length := len(q)
		for i := 0; i < length; i++ {
			point := q[0]
			q = q[1:]
			for _, direction := range directions {
				x, y := point[0]+direction[0], point[1]+direction[1]
				if inArea(x, y) && maze[x][y] == '.' {
					if x == 0 || y == 0 || x == m-1 || y == n-1 {
						return res
					}
					maze[x][y] = '+'
					q = append(q, [2]int{x, y})
				}
			}
		}
	}
	// 还没找到，置为 -1 返回
	return -1
}
// @lc code=end

File: 1926.迷宫中离入口最近的出口.py
#
# @lc app=leetcode.cn id=1926 lang=python3
#
# [1926] 迷宫中离入口最近的出口
#

# @lc code=start
class Solution:
    def nearestExit(self, maze: List[List[str]], entrance: List[int]) -> int:
        m, n = len(maze), len(maze[0])
        q = [(entrance[0], entrance[1])]
        directs = [(1, 0), (-1, 0), (0, 1), (0, -1)]

        def in_area(x, y):
            if 0 <= x < m and 0 <= y < n:
                return True
            return False

        maze[entrance[0]][entrance[1]] = "+"    # 起点已访问
        res = 1
        while q:
            for _ in range(len(q)):
                point = q[0]
                q = q[1:]
                for direct in directs:
                    x, y = point[0] + direct[0], point[1] + direct[1]
                    if in_area(x, y) and maze[x][y] == '.':
                        if x == 0 or y == 0 or x == m - 1 or y == n - 1:
                            return res
                        maze[x][y] = '+'    # 已访问
                        q.append((x, y))
            res += 1
        return -1
# @lc code=end
